README.md
# Introduction

[![Website](https://img.shields.io/badge/Visit%20Our-Website-blue)](https://nehemie-textguard.vercel.app)

Welcome to the @nehemie/textguard documentation!

### What is @6nehemie/textguard?

**@6nehemie/textguard** is a comprehensive JavaScript/TypeScript library engineered for robust data validation. This library allows you to enhance the integrity of various data inputs in your application, from emails and usernames to custom fields.

With its straightforward set of functions and methods, @6nehemie/textguard employs a range of best practices and algorithms to make sure your validation is both secure and efficient. This enables you to concentrate on building your application, relieving you from the complexities of data validation.

Whether you're an individual developer aiming to improve the security and reliability of your project, or part of a larger team in search of a reliable validation solution, @6nehemie/textguard is designed to meet your requirements.

### Features

- **Data Validation**: Robust validation for email, username, and password fields.
- **Custom Error Messages**: Provides customizable error messages for better user experience.
- **Easy Integration**: Designed to be easily integrated into existing projects.
- **TypeScript Support**: Comes with built-in TypeScript definitions.
- **Efficient**: Lightweight with minimal dependencies, optimized for performance.
- **Security**: Uses best practices to ensure data validation is secure and reliable.

## Installation

System Requirements:
Before installing @6nehemie/textguard, please ensure that your system meets the following requirements:

- Node.js: 16.14 or above. Note that this library uses ES modules, so make sure your Node.js environment supports it.
- TypeScript: Version 5.2.2 or above if you are using TypeScript in your project.

### Quick and Easy Installation Steps

**Using npm:**

To install @nehemie/textgurad via npm, run the following command:

```linux
npm i @6nehemie/textguard
```

**Using yarn:**

To install @nehemie/textgurad via yarn, run the following command:

```linux
yarn add @6nehemie/textguard
```

After installation, you are ready to import @6nehemie/textguard into your project and start laveragin its powerful features.

## Quick Start

**TypeScript**

### Example Usage

Here's a simple example that demonstrates email validation:

```typescript
import { isValidEmail } from '@6nehemie/textguard';

const email = 'example@example.com';
const validEmail = isValidEmail(email); // returns true
```

And that's it! You've successfully validated an email address.

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

### License

MIT


CHANGELOG.md
# @nemi-typescript/textguard

## 0.0.7

### Patch Changes

- 7b30d6f: New function and updated readme

## 0.0.6

### Patch Changes

- d2ce545: Including major function for data validation

## 0.0.5

### Patch Changes

- 587be79: valid email validator

## 0.0.4

### Patch Changes

- bb77c6c: Email validator

## 0.0.3

### Patch Changes

- 0a171cf: Removing the files that should not be sent

## 0.0.2

### Patch Changes

- b181140: Initialization

## 0.0.2

### Patch Changes

- da118ed: Initial

## 0.0.5

### Patch Changes

- 776d137: from github

## 0.0.4

### Patch Changes

- 6462e69: Initial version

## 0.0.5

### Patch Changes

- 224f5dc: initialization

## 0.0.4

### Patch Changes

- 7e04789: initial

## 0.0.3

### Patch Changes

- d3a5e1d: initial

## 0.0.2

### Patch Changes

- db109ab: initial

## 0.0.6

### Patch Changes

- 66ff517: initial

## 0.0.5

### Patch Changes

- 45f2807: initial commit

## 0.0.4

### Patch Changes

- f5e2b91: initial commit

## 0.0.3

### Patch Changes

- fd7e68c: Initial commit

## 0.0.2

### Patch Changes

- 46784b7: Initial commit


tests/index.test.ts
import { describe, it, expect } from 'vitest';

describe('Whatever', () => {
  it('should pass CI', () => {
    expect(1).toBe(1);
  });
});


tests/validators/stringValidators.test.ts
import { expect, describe, it } from 'vitest';
import {
  validateEmail,
  validateUsername,
  validatePassword,
  validateIsOldEnough,
} from '../../src';
import { ValidationError } from '../../src/errors/validationErrors';

describe('validateEmail', () => {
  it('should validate a correct email', () => {
    expect(() => validateEmail('test@test.com')).not.toThrow();
  });

  it('should throw an ValidationError for an invalid email', () => {
    expect(() => validateEmail('not an email address')).to.throw(
      'Invalid email address.'
    );
    expect(() => validateEmail('')).to.throw('Invalid email address.');
    expect(() => validateEmail('not an email address')).to.throw(
      'Invalid email address.'
    );
  });

  it('should throw an ValidationError for a wrongly typed email', () => {
    expect(() => validateEmail('testtest.com')).to.throw(
      'Invalid email address.'
    );
    expect(() => validateEmail('test@test.c')).to.throw(
      'Invalid email address.'
    );
    expect(() => validateEmail('test@testc')).to.throw(
      'Invalid email address.'
    );
    expect(() => validateEmail('test@test.444')).to.throw(
      'Invalid email address.'
    );
  });
});

describe('validateUsername', () => {
  it('shoud validate the username', () => {
    expect(() => validateUsername('username')).not.toThrow();
  });

  it('should throw an ValidationError for an invalid username', () => {
    expect(() => validateUsername('Username')).toThrow(
      'Username must start with a lowercase letter, contain only letters, numbers, spaces, periods, hyphens, or underscores, and be between 3 to 15 characters long.'
    );
    expect(() => validateUsername('')).toThrow(
      'Username must start with a lowercase letter, contain only letters, numbers, spaces, periods, hyphens, or underscores, and be between 3 to 15 characters long.'
    );
    expect(() => validateUsername('.@fiefnefe44')).toThrow(
      'Username must start with a lowercase letter, contain only letters, numbers, spaces, periods, hyphens, or underscores, and be between 3 to 15 characters long.'
    );
    expect(() => validateUsername('Username')).toThrow(
      'Username must start with a lowercase letter, contain only letters, numbers, spaces, periods, hyphens, or underscores, and be between 3 to 15 characters long.'
    );
  });
});

describe('validatePassword', () => {
  it('shoud validate the password', () => {
    expect(() => validatePassword('StrongPassword92.')).not.toThrow();
  });

  it('should throw an ValidationError for an invalid password', () => {
    expect(() => validatePassword('password')).toThrow(
      'Password must contain at least one uppercase letter, one lowercase letter, one digit, one special character (allowed: @$!%*?&.), and be at least 8 characters long.'
    );
    expect(() => validatePassword('')).toThrow(
      'Password must contain at least one uppercase letter, one lowercase letter, one digit, one special character (allowed: @$!%*?&.), and be at least 8 characters long.'
    );
    expect(() => validatePassword('missingSpecialCharacter')).toThrow(
      'Password must contain at least one uppercase letter, one lowercase letter, one digit, one special character (allowed: @$!%*?&.), and be at least 8 characters long.'
    );
  });
});

describe('validateIsOldEnough', () => {
  it('should validate the dates', () => {
    expect(() => validateIsOldEnough(18, 6, 3, 2000)).not.toThrow();
  });

  it('should throw an validation error for invalid dates', () => {
    expect(() => validateIsOldEnough(18, 34, 2, 2000)).toThrow();
    expect(() => validateIsOldEnough(18, 24, 0, 2000)).toThrow();
    expect(() => validateIsOldEnough(18, 24, -1, 2000)).toThrow();
    expect(() => validateIsOldEnough(18, 24, 54, 2000)).toThrow();
  });
});


tests/composite/compositeValidator.test.ts
import { expect, describe, it } from 'vitest';
import {
  isValidEmail,
  isValidUrsername,
  isValidPassword,
  arePasswordsMatching,
  isOldEnough,
} from '../../src';

describe('isValidEmail', () => {
  it('should return true if the email is valid', () => {
    expect(isValidEmail('test@test.com')).toBe(true);
  });

  it('should return false if the email is not valid', () => {
    expect(isValidEmail('not an email')).toBe(false);
    expect(isValidEmail('test@test.c')).toBe(false);
    expect(isValidEmail('test@test.44')).toBe(false);
    expect(isValidEmail('testtest.com')).toBe(false);
    expect(isValidEmail('test@testcom')).toBe(false);
  });
});

describe('isValidUrsername', () => {
  it('should return true if the username is valid', () => {
    expect(isValidUrsername('username')).toBe(true);
  });

  it('should return false if the username is not valid', () => {
    expect(isValidUrsername('')).toBe(false);
    expect(isValidUrsername('Username')).toBe(false);
    expect(isValidUrsername('@username')).toBe(false);
    expect(isValidUrsername('username%')).toBe(false);
  });
});

describe('isValidPassword', () => {
  it('should return true if the password is valid', () => {
    expect(isValidPassword('ValidPassword94.')).toBe(true);
  });
  it('should return false if the password is not valid', () => {
    expect(isValidPassword('NotaValidPassword')).toBe(false);
    expect(isValidPassword('notavalidpassword')).toBe(false);
    expect(isValidPassword('notavalidpassword8!')).toBe(false);
    expect(isValidPassword('NotaValidPassword1')).toBe(false);
  });
});

describe('arePasswordsMatching', () => {
  it('should return true if the password are matching', () => {
    expect(arePasswordsMatching('ValidPassword94.', 'ValidPassword94.')).toBe(
      true
    );
  });

  it('should return false if the passwords are not matching', () => {
    expect(arePasswordsMatching('ValidPassword94.', 'ValidPassword94')).toBe(
      false
    );
  });
});

describe('isOldEnough', () => {
  it('should return true if the user is old enough', () => {
    expect(isOldEnough(18, 3, 12, 2000)).toBe(true);
  });

  it('should return false if the user is not old enough', () => {
    expect(isOldEnough(18, 3, 12, 2020)).toBe(false);
  });
});


src/composite/compositeValidator.ts
import {
  validateEmail,
  validateUsername,
  validatePassword,
  validateIsOldEnough,
} from '../validators/stringValidators';
import { ValidationError } from '../errors/validationErrors';

/**
 * Validates the given email address based on a predefined regular expression pattern.
 * Throws a ValidationError if the email address is not valid.
 *
 * @param {string} email - The email address to validate.
 * @throws {ValidationError} Throws an error if the email is not valid.
 * @returns {boolean} Returns true if the email is valid, false otherwise.
 */
export const isValidEmail = (email: string): boolean => {
  try {
    validateEmail(email);
    return true;
  } catch (error) {
    if (error instanceof ValidationError) {
      return false;
    }
    throw error;
  }
};

/**
 * Validates the given username based on a predefined regular expression pattern.
 * Throws a ValidationError if the username is not valid.
 *
 * @param {string} username - The username to validate.
 * @throws {ValidationError} Throws an error if the username is not valid.
 * @returns {boolean} Returns true if the username is valid, false otherwise.
 */
export const isValidUrsername = (username: string): boolean => {
  try {
    validateUsername(username);
    return true;
  } catch (error) {
    if (error instanceof ValidationError) {
      return false;
    }
    throw error;
  }
};

/**
 * Validates the given password based on a predefined regular expression pattern.
 * Throws a ValidationError if the password is not valid.
 *
 * @param {string} password - The password to validate.
 * @throws {ValidationError} Throws an error if the password is not valid.
 * @returns {boolean} Returns true if the password is valid, false otherwise.
 */
export const isValidPassword = (password: string): boolean => {
  try {
    validatePassword(password);
    return true;
  } catch (error) {
    if (error instanceof ValidationError) {
      return false;
    }
    throw error;
  }
};

/**
 * Validates fi the given passwords are the same and  valid
 *
 * @param {string} password - The password to validate.
 * @param {string} confirmPassword - The password confirmation to validate.
 * @returns {boolean} Returns true if the passwords are the same, false otherwise.
 */
export const arePasswordsMatching = (
  password: string,
  confirmPassword: string
): boolean => {
  if (password !== confirmPassword) {
    return false;
  }

  return true;
};

/**
 * Determines if a person with the given birth date is old enough based on a minimum age requirement.
 *
 * @param {number} minAge - The minimum age the person must be.
 * @param {number} day - The day of the person's birth date (1-31).
 * @param {number} month - The month of the person's birth date (1-12).
 * @param {number} year - The year of the person's birth date.
 * @returns {boolean} Returns true if the person is at least the specified minAge or older, false otherwise.
 */
export const isOldEnough = (
  minAge: number,
  day: number,
  month: number,
  year: number
): boolean => {
  validateIsOldEnough(minAge, day, month, year);
  const currentDate = new Date();
  const birthDate = new Date(year, month - 1, day);

  let age = currentDate.getFullYear() - birthDate.getFullYear();

  if (
    currentDate.getMonth() < month - 1 ||
    (currentDate.getMonth() === month - 1 && currentDate.getDate() < day)
  ) {
    age--;
  }

  if (age >= minAge) {
    return true;
  } else {
    return false;
  }
};


src/validators/stringValidators.ts
import { ValidationError } from '../errors/validationErrors';

/**
 * Validates an email address against a predefined regular expression.
 *
 * @param {string} email - The email address to validate.
 * @throws {ValidationError} - Throws an error if the email address is not valid.
 * @returns {void} - Returns nothing if the email address is valid.
 */
export const validateEmail = (email: string): void => {
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;

  if (!emailRegex.test(email)) {
    throw new ValidationError('Invalid email address.');
  }
};

/**
 * Validates a username against a predefined regular expression.
 *
 * @param {string} username - The username to validate.
 * @throws {ValidationError} - Throws an error if the username is not valid.
 * @returns {void} - Returns nothing if the username address is valid.
 */
export const validateUsername = (username: string): void => {
  const usernameRegex = /^[a-z][a-z0-9 .-_]{2,14}$/;

  if (!usernameRegex.test(username)) {
    throw new ValidationError(
      'Username must start with a lowercase letter, contain only letters, numbers, spaces, periods, hyphens, or underscores, and be between 3 to 15 characters long.'
    );
  }
};

/**
 * Validates a password against a predefined regular expression.
 *
 * @param {string} password - The password to validate.
 * @throws {ValidationError} - Throws an error if the password is not valid.
 * @returns {void} - Returns nothing if the password address is valid.
 */
export const validatePassword = (password: string): void => {
  const passwordRegex =
    /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&.])[A-Za-z\d@$!%*?&.]{8,}$/;

  if (!passwordRegex.test(password)) {
    throw new ValidationError(
      'Password must contain at least one uppercase letter, one lowercase letter, one digit, one special character (allowed: @$!%*?&.), and be at least 8 characters long.'
    );
  }
};

/**
 * Validates the provided birth date details and ensures they form a valid date.
 * Also checks that the 'minAge' parameter is a valid number.
 *
 * This function will throw a ValidationError if:
 * 1. Any of the parameters are not of type 'number'.
 * 2. The month value is not between 1 and 12.
 * 3. The day value is not valid for the given month (e.g., February 30th).
 *
 * @param {number} minAge - The minimum age value to validate.
 * @param {number} day - The day of the birth date to validate (1-31, depending on month).
 * @param {number} month - The month of the birth date to validate (1-12).
 * @param {number} year - The year of the birth date to validate.
 * @throws {ValidationError} Throws error if the input values are invalid.
 */
export const validateIsOldEnough = (
  minAge: number,
  day: number,
  month: number,
  year: number
) => {
  if (
    typeof minAge !== 'number' ||
    typeof day !== 'number' ||
    typeof month !== 'number' ||
    typeof year !== 'number'
  ) {
    throw new ValidationError('Dates must only contain numbers.');
  }

  if (month < 1 || month > 12) {
    throw new ValidationError('Month must be between 1 and 12.');
  }

  let daysInMonth;
  switch (month) {
    case 2: // February
      // Check for leap year
      if ((year % 4 === 0 && year % 100 !== 0) || year % 400 === 0) {
        daysInMonth = 29;
      } else {
        daysInMonth = 28;
      }
      break;
    case 4: // April
    case 6: // June
    case 9: // September
    case 11: // November
      daysInMonth = 30;
      break;
    default:
      daysInMonth = 31;
  }

  if (day < 1 || day > daysInMonth) {
    throw new ValidationError(
      `Day must be between 1 and ${daysInMonth} for the given month.`
    );
  }
};
